File: p125e/valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        lo, hi = 0, len(s) - 1
        while lo < hi:
            if not s[lo].isalnum():
                lo += 1
            elif not s[hi].isalnum():
                hi -= 1
            else:
                if s[lo].lower() != s[hi].lower():
                    return False
                lo, hi = lo + 1, hi - 1
        return True


# TESTS
tests = [
    ["", True],
    ["A man, a plan, a canal: Panama", True],
    ["race a car", False],
    ["0P", False],
]
for t in tests:
    sol = Solution()
    print('Is "' + t[0] + '" palindrome? ->', t[1])
    assert sol.isPalindrome(t[0]) == t[1]

File: p125e/valid_palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
  let [lo, hi] = [0, s.length - 1];
  while (lo < hi) {
    if (s[lo].match(/[0-9a-zA-Z]/) === null) {
      lo++;
    } else if (s[hi].match(/[0-9a-zA-Z]/) === null) {
      hi--;
    } else {
      if (s[lo].toLowerCase() !== s[hi].toLowerCase()) return false;
      lo++;
      hi--;
    }
  }
  return true;
};
[
  ['', true],
  ['A man, a plan, a canal: Panama', true],
  ['race a car', false],
  ['0P', false]
].forEach(t => {
  let act = isPalindrome(t[0]);
  console.log('Is "' + t[0] + '" palindrome? ->', act);
  console.assert(act === t[1]);
});
